app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt

```diff
@@ -1,32 +1,97 @@
+import io.reactivex.rxjava3.disposables.Disposable
 
 /**
  * This view model manages the data for [FeedFragment].
  */
-open class FeedViewModel : ViewModel() {
-    private val isLoading = MutableLiveData<Boolean>()
-    private val isEmpty = MutableLiveData<Boolean>()
-    private val feedItems = MediatorLiveData<List<FeedItem>>()
+open class FeedViewModel(private val feedRepository: FeedRepository) : ViewModel() {
+
+    private val stateLiveData = MutableLiveData<State>()
     private val networkErrorEvent = MutableLiveData<Event<String>>()
+    private var prevTask: Disposable? = null
+    private lateinit var getFeedFromRepoStream: Disposable
+
+    fun getIsLoading(): LiveData<Boolean> = Transformations.map(stateLiveData) { it.isLoading }
+    fun getIsEmpty(): LiveData<Boolean> = Transformations.map(stateLiveData) { it.isFeedEmpty() }
+    fun getFeedItems(): LiveData<List<FeedItem>> =
+        Transformations.map(stateLiveData) { it.feedItems }
 
-    fun getIsLoading(): LiveData<Boolean> = isLoading
-    fun getIsEmpty(): LiveData<Boolean> = isEmpty
-    fun getFeedItems(): LiveData<List<FeedItem>> = feedItems
     fun getNetworkErrorEvent(): LiveData<Event<String>> = networkErrorEvent
 
     init {
+        observeRepository()
         refresh()
     }
 
     fun refresh() {
-        //todo: fix the implementation
-        isLoading.value = false
-        isEmpty.value = true
+        setStartLoading()
+        disposePreviousTask()
+        prevTask = feedRepository.fetchFeed()
+            .observeOn(AndroidSchedulers.mainThread())
+            .subscribe({ setStoppedLoading() },
+                { error ->
+                    handleNetworkError(error)
+                    setStoppedLoading()
+                })
+    }
+
+    private fun disposePreviousTask() {
+        prevTask?.dispose()
+    }
+
+    private fun handleNetworkError(error: Throwable?) {
+        networkErrorEvent.value = Event(R.string.repository_network_error.toString())
+        Log.e(TAG, "Failed to fetch feed from network", error)
+    }
+
+    private fun observeRepository() {
+        getFeedFromRepoStream = feedRepository.getFeedItems().subscribe({ items ->
+            stateLiveData.value = when (stateLiveData.value) {
+                null -> State(feedItems = items)
+                else -> stateLiveData.value!!.copy(feedItems = items)
+            }
+        }, { error -> Log.e(TAG, "Failed to get feed from repository", error) })
+    }
+
+    private fun setStartLoading() {
+        stateLiveData.value?.isLoading = true
+    }
+
+    private fun setStoppedLoading() {
+        stateLiveData.value?.isLoading = false
+    }
+
+    data class State(
+        var isLoading: Boolean = false,
+        val feedItems: List<FeedItem> = emptyList()
+    ) {
+        fun isFeedEmpty(): Boolean {
+            return feedItems.isEmpty()
+        }
+    }
+
+    override fun onCleared() {
+        super.onCleared()
```

Comment:
super.onCleared does nothing so you can remove it.

Reply:
I think it is logically necessary to keep this line, since we expand the function functionality. Even if in this specific case the implementation of super.onCleared() does nothing currently.